hane-kernel/src/stack.rs

```diff
@@ -1,88 +1,135 @@
 use std::iter::Rev;
 use std::mem::ManuallyDrop;
+use std::ops::{Deref, DerefMut};
 
-/// A stack separated into sections called slots. Each `Stack<'a, T>` represents one such slot.
-/// When a slot is droped so is all the elements of that section.
-pub struct Stack<'a, T> {
-    buf: ManuallyDrop<&'a mut Vec<T>>,
-    /// The starting index of this stack slot
-    slot: usize,
-}
+/// First in last out stack, with indexation from newest to oldest entry.
+/// All elements must be inserted through a `StackSlot`.
+#[derive(Default, Clone)]
+pub struct Stack<T>(Vec<T>);
 
-impl<'a, T> Drop for Stack<'a, T> {
-    fn drop(&mut self) {
-        self.buf.truncate(self.slot);
+impl<T> From<Vec<T>> for Stack<T> {
+    fn from(buf: Vec<T>) -> Self {
+        Stack(buf)
     }
 }
 
-impl<'a, T> Stack<'a, T> {
-    pub fn new(buf: &'a mut Vec<T>) -> Self {
-        Stack {
-            slot: buf.len(),
-            buf: ManuallyDrop::new(buf),
-        }
+impl<T> Stack<T> {
+    pub fn new() -> Self {
+        Stack(Vec::new())
     }
 
+    /// Returns the number of elements in the stack, also referred to as its 'length'.
     pub fn len(&self) -> usize {
-        self.buf.len()
+        self.0.len()
     }
 
+    /// Returns true if the vector contains no elements.
     pub fn is_empty(&self) -> bool {
-        self.buf.is_empty()
+        self.0.is_empty()
+    }
+
+    /// Returns a reference to the `index`th newest element or None if out of bounds.
+    pub fn get(&self, index: usize) -> Option<&T> {
+        if index < self.0.len() {
+            // Safety: `index` is checked to be within bounds
+            Some(unsafe { self.0.get_unchecked(self.0.len() - 1 - index) })
+        } else {
+            None
+        }
+    }
+
+    /// Shortens the stack, keeping the oldest len elements and dropping the rest.
+    /// If len is greater than the stack's current length, this has no effect.
+    pub fn truncate(&mut self, len: usize) {
+        self.0.truncate(len)
+    }
+
+    /// Returns true if the stack contains an element with the given value.
+    /// The iterator yields all items from newest to oldest.
+    pub fn contains(&self, x: &T) -> bool
+    where
+        T: PartialEq,
+    {
+        self.0.contains(x)
     }
 
-    /// Creates an empty stack slot
-    #[must_use]
-    pub fn new_slot<'b>(&'b mut self) -> Stack<'b, T> {
-        Stack::new(&mut *self.buf)
+    /// Returns an iterator over the stack.
+    pub fn iter(&self) -> Rev<std::slice::Iter<T>> {
```

Comment:
I still think it's better to promise a minimal API (`impl Iterator<Item = T> + '_`), than breaking our callers e.g. if we stop using `std::slice::Iter` underneath.